Curl security schemes

hotgazpacho opened the pull request:
Fixes #28 by introducing processing of the `securedBy` attribute of a given method.
This takes into account the specific rules for each of the RAML-defined security scheme types, including multiple schemes and explicit null schemes. For multiple security schemes, or for OAuth 2 _(where the spec says you can pass either a header or a query string parameter)_, multiple curl commands are created.

hotgazpacho commented:
The one scheme I have not implemented is OAuth 1.0. I haven't read up enough yet on the protocol to understand exactly which headers and/or query parameters need to be set for a given request.

hotgazpacho commented:
Alright, added support for OAuth 1.

Question: In the case where multiple security schemes are defined for a method, do we want to show multiple curl commands for the same method, or should we simply pick the first one, presuming that it will be the preferred scheme? What about the case of OAuth (1 or 2), where the credentials can be passed via header or query string?

wdullaer commented on lib/stylus-globals.js:
```diff
@@ -54,17 +55,193 @@ function getCurlStatement (baseUri, method, resource) {
+      if (methodScheme) {
+        let securityScheme = securitySchemes[methodScheme.schemeName]
+        if (securityScheme) {
+          Array.prototype.push.apply(commands, buildCurlCommands(resourceOptions, securityScheme))
```
wouldn't 
```javascript
commands = commands.concat(buildCurlCommands(resourceOptions, securityScheme))
```
do the same?

hotgazpacho replied:
If I changed `const commands` to `let commands` on line 75, yes.

wdullaer commented on lib/stylus-globals.js:
```diff
@@ -54,17 +55,193 @@ function getCurlStatement (baseUri, method, resource) {
+          Array.prototype.push.apply(commands, buildCurlCommands(resourceOptions, securityScheme))
+        }
+      } else {
+        Array.prototype.push.apply(commands, buildCurlCommands(resourceOptions, {type: 'null'}))
```
I would add `{type: 'null'}` in `buildCurlCommands` as a default value.

wdullaer commented on lib/stylus-globals.js:
```diff
@@ -54,17 +55,193 @@ function getCurlStatement (baseUri, method, resource) {
+  const commands = []
+  if (securitySchemes && method.securedBy) {
+    method.securedBy.forEach((methodScheme) => {
+      if (methodScheme) {
```
this could be
```javascript
if (methodScheme && securitySchemes[methodScheme.schemeName])
```
That way you get rid of one of the else blocks (both elses seem to do the same)

hotgazpacho replied:
Hmmm. One of the `else`s is nested inside a `forEach` loop, which loops over all the security schemes for the method. That inner else guards against the scenario where the method has named a security scheme, but a matching security scheme hasn't been defined at the root. Of course, this would be an invalid RAML document, but I figured we should guard against it.

hotgazpacho replied:
It also handles the case where multiple security schemes are defined on the method, and one of them is explicitly `null`. The test case for this is on line 433 of stylus-globals.test.js.

wdullaer commented on lib/stylus-globals.js:
```diff
@@ -54,17 +55,193 @@ function getCurlStatement (baseUri, method, resource) {
   const relativeUri = resource.relativeUri || '/'
 
-  const headers = method.headers
+  let headers = method.headers
```
Just by reading the code here on github, I think this can still be a `const`

wdullaer commented on lib/stylus-globals.js:
```diff
@@ -54,17 +55,193 @@ function getCurlStatement (baseUri, method, resource) {
+ * @param  {string}   methodParams.method  The HTTP method
+ * @param  {string}   methodParams.baseUri The base URI
+ * @param  {string}   methodParams.path    The path
+ * @param  {String[]} methodParams.params  An array of serialized key-value pairs that make up the query string parameters
```
I know it's a massive nitpick but I mostly use `array<string>` in my code.

wdullaer commented on lib/stylus-globals.js:
```diff
@@ -54,17 +55,193 @@ function getCurlStatement (baseUri, method, resource) {
+  }
+
+  let authFunction = curlAuthFunctions[securityScheme.type]
+  if (authFunction === undefined) { authFunction = curlNullAuth }
```
these two lines can be collapsed to
```javascript
const authFunction = curlAuthFunction[securityScheme.type] || curlNullAuth
```

hotgazpacho commented:
> In general it's really nice work!

Thanks!

> I would have just picked the first security scheme and rendered that one.

I can definitely do that.

> I'm also thinking that maybe all the code for rendering the curl commands should go into a separate file. This one is getting rather large now.

Agree completely! I'll get on that.

hotgazpacho commented:
Making progress on extracting the `curl-command-builder` module. Refactoring a bit along the way, with the goal of moving the `getCurlStatement` tests over.

hotgazpacho commented:
OK, I believe I have addressed all the feedback, including extracting the curl command builder to a separate module, and making clear the logic in the `buildSecure` method.

hotgazpacho commented:
> would add {type: 'null'} in buildCurlCommands as a default value.

That seems to be making the build fail on Node 4. Do you still want to support Node 4? If so, then I think I need to move this back inside the body of the function.

hotgazpacho commented:
@wdullaer any thoughts on my changes and the build failing for Node 4?

wdullaer commented:
I have been traveling for work and didn't really have time to review.
I'll try to get around to it today / tomorrow

wdullaer commented on lib/curl-command-builder.js:
```diff
@@ -0,0 +1,221 @@
+ * @param  {string}        methodParams.baseUri   The base URI
+ * @param  {string}        methodParams.path      The path
+ * @param  {array<string>} methodParams.params    An array of serialized key-value pairs that make up the query string parameters
+ * @param  {array<string>} methodParams.headers   An array of sewrialized HTTP headers
```
minor typo here

wdullaer commented on lib/curl-command-builder.js:
```diff
@@ -0,0 +1,221 @@
+ * @param  {object}        securityScheme       A Security Scheme object as generated by RAML
+ * @return {array<string>}                      An array of strings that represents the possible curl commands
+ */
+function buildCurlCommands (methodParams, securityScheme = {type: 'null'}) {
```
Default parameters in the argument definition don't work on node < 6.
What I often do is put something like this in the function body to set a default value
```js
securityScheme = securityScheme || {type: 'null'}
```

hotgazpacho commented:
> This looks great. I really appreciate the effort you've put in!

My pleasure! I'd be lying if I said I wasn't scratching my own itch here 😄 

I've fixed the typo, and moved the default parameter initialization to inside the function body.

hotgazpacho commented:
Just ran this against a RAML project I'm working on, and realized that I want to potentially include the examples for the headers or params, if provided. Will work on that tonight.

hotgazpacho commented:
OK, everything seems green now, and I've addressed the review comments 😄

wdullaer commented:
The code looks good to me.
I just want to try it out on some example raml schemas I've used to develop this. I should have some time for that later this evening.

wdullaer commented on lib/curl-command-builder.js:
```diff
@@ -0,0 +1,236 @@
+ */
+function curlXCustomAuth (securityScheme) {
+  return [{
+    headers: securityScheme.describedBy.headers.map((header) => `-H "${header.name}: ${header.type}"`)
```
Maybe also do `exampleOrDefault(header)` here instead of `header.type`?

wdullaer commented on lib/curl-command-builder.js:
```diff
@@ -0,0 +1,236 @@
+  }
+
+  securityScheme = securityScheme || {type: 'null'}
+  const authFunction = curlAuthFunctions[securityScheme.type] || curlNullAuth
```
The spec allows for `x-{other}` as a custom type, so `x-custom` is not sufficient.
I suppose we could do a pipeline that shortcircuits:
```js
const authFunction = curlAuthFunctions[securityScheme.type] || curlCustomAuth(securityScheme.type) || curlNullAuth

function curlCustomAuth(type) {
  return type.startsWith('x-') ? curlXCustomAuth : false
}
```
Or maybe wrap the pipeline into a single function

hotgazpacho replied:
Oh, interesting. I read the spec as saying that the value for `type` will be `x-custom`: https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md#x-other

wdullaer commented:
Great!
I'll try to release this later today.

On a side note: would you be interested in helping me maintain this theme?

hotgazpacho commented:
> On a side note: would you be interested in helping me maintain this theme?

Sure! I have some thoughts on exposing additional things, like the security schemes... 😈

wdullaer commented:
Cool.
Now I just need to figure out how to give you commit rights :-)
